### ISSUE TITLE
Strange behavior with DPI scaling

### ISSUE THREAD

**Issue by sudoforge** [alacritty/alacritty#4797]
### System

OS: Linux
Version: `alacritty 0.7.1 (a852eb85)`
Linux/BSD: X11/Wayland, DE or Compositor and WM

```
➜ pacman -Qi xorg-server bspwm lightdm | rg -B 1 \^Version
Name            : xorg-server
Version         : 1.20.10-3
--
Name            : bspwm
Version         : 0.9.10-1
--
Name            : lightdm
Version         : 1:1.30.0-4
```

### Description

I have two displays; only one of which is active at any given time. The first "small" display (`eDP-1`) is the inbuilt 14 inch, 1920x1080 display on my laptop. The second (`DP-2`) is a 27 inch, 2560x1440 external display (Dell U2721DE). When I connect my laptop to the monitor, I have `autorandr` perform the following tasks:

* Turn `eDP-1` off, and set `DP-2` as the primary
* Move bspwm desktops to the new display (meaning windows like alacritty get relocated and have more real estate available)
* Restart a systemd userspace service for `polybar`, so that the bar is moved to the appropriate display

When disconnecting, the inverse of step one happens.

When Alacritty is spawned on `eDP-1`, with a `font.size` of `6`, the visual representation of text on screen is satisfactory. When connecting `DP-2`, the font size is scaled somewhat appropriately. *The issue, however* is twofold:

* When spawning a new Alacritty window on `DP-2`, no scaling occurs and my `font.size` of `6` results in a font display size that is too small to comfortably read.
* With that same window, above, _disconnecting_ from `DP-2` (and moving back to `eDP-1`) *does not* scale the text to an appropriate size -- it appears as if Alacritty thinks my display size and DPI have not changed (this is only fixed by closing Alacritty and reopening on `eDP-1`).

My `alacritty.yml` configuration is fairly light-touch, relying on the default value for many settings, and can be seen here:

https://github.com/sudoforge/dotfiles/blob/trunk/alacritty/.config/alacritty/alacritty.yml

### Logs

A fresh window on `eDP-1`:

```
➜ alacritty -vv
Created log file at "/tmp/Alacritty-50354.log"
[2021-02-15 20:07:24.648747372] [INFO ] [alacritty] Welcome to Alacritty
[2021-02-15 20:07:24.648835236] [INFO ] [alacritty] Configuration files loaded from:
                                                      "/home/sudoforge/.config/alacritty/alacritty.yml"
[2021-02-15 20:07:24.665272784] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:07:24.668672325] [DEBUG] [alacritty] Estimated DPR: 1.6666666666666667
[2021-02-15 20:07:24.668708406] [DEBUG] [alacritty] Estimated window size: None
[2021-02-15 20:07:24.668719878] [DEBUG] [alacritty] Estimated cell size: 8 x 16
[2021-02-15 20:07:24.789380175] [INFO ] [alacritty] Device pixel ratio: 1.6666666666666667
[2021-02-15 20:07:24.800951607] [INFO ] [alacritty] Initializing glyph cache...
[2021-02-15 20:07:24.808727137] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:07:24.821166963] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:07:24.833556581] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: NO_BITMAP | TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:07:24.846224628] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: NO_BITMAP | TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:07:24.872883293] [INFO ] [alacritty] ... finished initializing glyph cache in 0.071869642s
[2021-02-15 20:07:24.873099750] [INFO ] [alacritty] Cell size: 8 x 16
[2021-02-15 20:07:24.873141573] [INFO ] [alacritty] Padding: 0 x 0
[2021-02-15 20:07:24.873159137] [INFO ] [alacritty] Width: 800, Height: 600
[2021-02-15 20:07:24.896859713] [INFO ] [alacritty] PTY dimensions: Line(37) x Column(100)
[2021-02-15 20:07:24.903016046] [INFO ] [alacritty] Initialisation complete
[2021-02-15 20:07:24.903745816] [DEBUG] [alacritty_terminal] New num_cols is 240 and num_lines is 66
[2021-02-15 20:07:24.909526337] [INFO ] [alacritty] Padding: 0 x 0
[2021-02-15 20:07:24.909586766] [INFO ] [alacritty] Width: 1920, Height: 1056
[2021-02-15 20:07:25.186907601] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:07:26.690592449] [INFO ] [alacritty] Goodbye
```

A fresh window on `DP-2`:

```
➜ alacritty -vv
Created log file at "/tmp/Alacritty-51887.log"
[2021-02-15 20:08:29.516273644] [INFO ] [alacritty] Welcome to Alacritty
[2021-02-15 20:08:29.516325491] [INFO ] [alacritty] Configuration files loaded from:
                                                      "/home/sudoforge/.config/alacritty/alacritty.yml"
[2021-02-15 20:08:29.529019575] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:08:29.531368525] [DEBUG] [alacritty] Estimated DPR: 1.1666666666666667
[2021-02-15 20:08:29.531391934] [DEBUG] [alacritty] Estimated window size: None
[2021-02-15 20:08:29.531401891] [DEBUG] [alacritty] Estimated cell size: 6 x 11
[2021-02-15 20:08:29.588983991] [INFO ] [alacritty] Device pixel ratio: 1.1666666666666667
[2021-02-15 20:08:29.594535904] [INFO ] [alacritty] Initializing glyph cache...
[2021-02-15 20:08:29.598279504] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:08:29.604043873] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:08:29.609243545] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: NO_BITMAP | TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:08:29.614236631] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: NO_BITMAP | TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:08:29.625694761] [INFO ] [alacritty] ... finished initializing glyph cache in 0.031118349s
[2021-02-15 20:08:29.625771390] [INFO ] [alacritty] Cell size: 6 x 11
[2021-02-15 20:08:29.625781416] [INFO ] [alacritty] Padding: 0 x 0
[2021-02-15 20:08:29.625788037] [INFO ] [alacritty] Width: 800, Height: 600
[2021-02-15 20:08:29.635551826] [INFO ] [alacritty] PTY dimensions: Line(54) x Column(133)
[2021-02-15 20:08:29.640177715] [INFO ] [alacritty] Initialisation complete
[2021-02-15 20:08:29.640745217] [DEBUG] [alacritty_terminal] New num_cols is 426 and num_lines is 128
[2021-02-15 20:08:29.644439124] [INFO ] [alacritty] Padding: 0 x 0
[2021-02-15 20:08:29.644475935] [INFO ] [alacritty] Width: 2560, Height: 1416
[2021-02-15 20:08:29.884879217] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:08:30.781370813] [INFO ] [alacritty] Goodbye
```

Spawned on `eDP-1`, moved to `DP-2`:

```
➜ alacritty -vv
Created log file at "/tmp/Alacritty-52745.log"
[2021-02-15 20:09:12.321359760] [INFO ] [alacritty] Welcome to Alacritty
[2021-02-15 20:09:12.321463626] [INFO ] [alacritty] Configuration files loaded from:
                                                      "/home/sudoforge/.config/alacritty/alacritty.yml"
[2021-02-15 20:09:12.347800789] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:09:12.353533015] [DEBUG] [alacritty] Estimated DPR: 1.6666666666666667
[2021-02-15 20:09:12.353599511] [DEBUG] [alacritty] Estimated window size: None
[2021-02-15 20:09:12.353625944] [DEBUG] [alacritty] Estimated cell size: 8 x 16
[2021-02-15 20:09:12.473589906] [INFO ] [alacritty] Device pixel ratio: 1.6666666666666667
[2021-02-15 20:09:12.480261822] [INFO ] [alacritty] Initializing glyph cache...
[2021-02-15 20:09:12.484825584] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:09:12.492180028] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:09:12.499438650] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: NO_BITMAP | TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:09:12.506669969] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: NO_BITMAP | TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:09:12.523237594] [INFO ] [alacritty] ... finished initializing glyph cache in 0.042926714s
[2021-02-15 20:09:12.523348361] [INFO ] [alacritty] Cell size: 8 x 16
[2021-02-15 20:09:12.523365608] [INFO ] [alacritty] Padding: 0 x 0
[2021-02-15 20:09:12.523374152] [INFO ] [alacritty] Width: 800, Height: 600
[2021-02-15 20:09:12.541357696] [INFO ] [alacritty] PTY dimensions: Line(37) x Column(100)
[2021-02-15 20:09:12.548591128] [INFO ] [alacritty] Initialisation complete
[2021-02-15 20:09:12.549326381] [DEBUG] [alacritty_terminal] New num_cols is 240 and num_lines is 66
[2021-02-15 20:09:12.554996878] [INFO ] [alacritty] Padding: 0 x 0
[2021-02-15 20:09:12.555066389] [INFO ] [alacritty] Width: 1920, Height: 1056
[2021-02-15 20:09:12.830952389] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:09:27.255069141] [DEBUG] [alacritty_terminal] New num_cols is 320 and num_lines is 90
[2021-02-15 20:09:27.263363248] [INFO ] [alacritty] Padding: 0 x 0
[2021-02-15 20:09:27.263431035] [INFO ] [alacritty] Width: 2560, Height: 1440
[2021-02-15 20:09:28.325694807] [DEBUG] [alacritty_terminal] New num_cols is 320 and num_lines is 88
[2021-02-15 20:09:28.325786863] [INFO ] [alacritty] Padding: 0 x 0
[2021-02-15 20:09:28.325801285] [INFO ] [alacritty] Width: 2560, Height: 1416
[2021-02-15 20:09:34.275462197] [INFO ] [alacritty] Goodbye
```

Spawned on `DP-2` (text is small), moved to `eDP-1` (DPI scale doesn't appear to change at all until respawning a window):

```
➜ alacritty -vv
Created log file at "/tmp/Alacritty-54398.log"
[2021-02-15 20:10:51.607034196] [INFO ] [alacritty] Welcome to Alacritty
[2021-02-15 20:10:51.607144478] [INFO ] [alacritty] Configuration files loaded from:
                                                      "/home/sudoforge/.config/alacritty/alacritty.yml"
[2021-02-15 20:10:51.626063465] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:10:51.628288122] [DEBUG] [alacritty] Estimated DPR: 1.1666666666666667
[2021-02-15 20:10:51.628310168] [DEBUG] [alacritty] Estimated window size: None
[2021-02-15 20:10:51.628317768] [DEBUG] [alacritty] Estimated cell size: 6 x 11
[2021-02-15 20:10:51.682945798] [INFO ] [alacritty] Device pixel ratio: 1.1666666666666667
[2021-02-15 20:10:51.688940374] [INFO ] [alacritty] Initializing glyph cache...
[2021-02-15 20:10:51.691763967] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:10:51.696870655] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:10:51.701539832] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: NO_BITMAP | TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:10:51.706601174] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: NO_BITMAP | TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:10:51.717385596] [INFO ] [alacritty] ... finished initializing glyph cache in 0.028414513s
[2021-02-15 20:10:51.717472660] [INFO ] [alacritty] Cell size: 6 x 11
[2021-02-15 20:10:51.717482635] [INFO ] [alacritty] Padding: 0 x 0
[2021-02-15 20:10:51.717488159] [INFO ] [alacritty] Width: 800, Height: 600
[2021-02-15 20:10:51.738000183] [INFO ] [alacritty] PTY dimensions: Line(54) x Column(133)
[2021-02-15 20:10:51.742927180] [INFO ] [alacritty] Initialisation complete
[2021-02-15 20:10:51.743304558] [DEBUG] [alacritty_terminal] New num_cols is 426 and num_lines is 128
[2021-02-15 20:10:51.746677550] [INFO ] [alacritty] Padding: 0 x 0
[2021-02-15 20:10:51.746719548] [INFO ] [alacritty] Width: 2560, Height: 1416
[2021-02-15 20:10:51.989302181] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:10:55.640415425] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:10:55.708833309] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:10:55.745262638] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: NO_BITMAP | TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:10:55.755399859] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: NO_BITMAP | TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:10:55.759087630] [INFO ] [alacritty] Font size changed to Size(12) with DPR of 1
[2021-02-15 20:10:55.762804916] [INFO ] [alacritty] Cell size: 5 x 10
[2021-02-15 20:10:55.762862601] [DEBUG] [alacritty_terminal] New num_cols is 438 and num_lines is 121
[2021-02-15 20:10:55.763849418] [INFO ] [alacritty] Padding: 0 x 0
[2021-02-15 20:10:55.763890591] [INFO ] [alacritty] Width: 2194, Height: 1214
[2021-02-15 20:10:55.765481490] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT | TARGET_MONO | TARGET_LCD, render_mode: "Lcd", lcd_filter: 1 }
[2021-02-15 20:10:55.793899575] [INFO ] [alacritty] Font size changed to Size(12) with DPR of 1
[2021-02-15 20:10:55.797152791] [INFO ] [alacritty] Cell size: 5 x 10
[2021-02-15 20:10:55.797219678] [DEBUG] [alacritty_terminal] New num_cols is 512 and num_lines is 141
[2021-02-15 20:10:55.807387916] [INFO ] [alacritty] Padding: 0 x 0
[2021-02-15 20:10:55.807421941] [INFO ] [alacritty] Width: 2560, Height: 1416
[2021-02-15 20:10:56.291729303] [DEBUG] [alacritty_terminal] New num_cols is 384 and num_lines is 108
[2021-02-15 20:10:56.293293781] [INFO ] [alacritty] Padding: 0 x 0
[2021-02-15 20:10:56.293340056] [INFO ] [alacritty] Width: 1920, Height: 1080
[2021-02-15 20:10:56.577642543] [DEBUG] [alacritty_terminal] New num_cols is 384 and num_lines is 105
[2021-02-15 20:10:56.577749058] [INFO ] [alacritty] Padding: 0 x 0
[2021-02-15 20:10:56.577767123] [INFO ] [alacritty] Width: 1920, Height: 1056
[2021-02-15 20:11:07.086535575] [INFO ] [alacritty] Goodbye
```

**Comment by chrisduerr** [comment]
Sounds to me like what you actually want is the font *not* scaling by DPI? You say the text is tiny on `DP-2`, but the DPR looks accurately lower.

**Comment by nixpulvis** [comment]
I'm trying to understand how it's possible that spawning on `eDP-1` and moving to `DP-2` doesn't cause a DPR update, but the reverse from `DP-2` to `eDP-1` does. Am I misreading these last two logs?

**Comment by sudoforge** [comment]
@chrisduerr 

> Sounds to me like what you actually want is the font not scaling by DPI? You say the text is tiny on DP-2, but the DPR looks accurately lower.

Possibly? I feel like I'm conflating the terms "scale", "DPI", and "DPR", along with how Alacritty _should_ be behaving upon display changes. I would imagine that a smaller DPR would result in scaling the text _up_, not down. Is that assumption incorrect?

Actually, in taking screenshots to showcase the issue, I noticed that there actually isn't any scaling when spawning a window on `eDP1` and moving to `DP-2`; as you can see from the sceenshots added to my initial comment, the text size is exactly the same (it appears visually larger to me because of the larger display, I imagine). This is the behavior I _want_. All other behavior appears to be incorrect, or at least, incorrect based upon my assumptions of what _should_ be happening.

@nixpulvis 

> Am I misreading these last two logs?

The only redacted line was the log file deletion (funnily enough, it doesn't appear that `--persistent-logging` is a valid flag anymore [although it's included in my completion suggestions], but I'll defer that discussion to another issue after looking into the cause).

That said, to your question specifically, I don't ever see that the DPR is updated (at least, no info message is provided, even with more verbose logging). In the `eDP-1` to `DP-2` workflow, as I described above, the text becomes visibly larger -- although this appears to be a visual trick caused by the larger display; nothing actually changes and the text is still the exact same size.

---

TL;DR: I added screenshots to my initial comment which can showcase the issue I'm experiencing.

**Comment by nixpulvis** [comment]
You say this is happening on **both** X11 and Wayland? I find it somewhat surprising the exact same behavior would be present for both of them.

**Comment by sudoforge** [comment]
> You say this is happening on both X11 and Wayland? I find it somewhat surprising the exact same behavior would be present for both of them.

I am not using Wayland and have not tested on Wayland.

**Comment by sudoforge** [comment]
Oh, I see where you got that from -- this line that was originally in my initial comment:

> Linux/BSD: X11/Wayland, DE or Compositor and WM

... is actually from the issue template. I forgot to clean that up when I decided to instead post the package versions.

**Comment by chrisduerr** [comment]
>  I would imagine that a smaller DPR would result in scaling the text up, not down. Is that assumption incorrect?

It is incorrect. Smaller -> Smaller, Bigger -> Bigger. The font size is scaled directly by DPI scale factor.

When you try running `WINIT_X11_SCALE_FACTOR=1.6 ./target/debug/alacritty`, does that behave more like you would want to?

**Comment by sudoforge** [comment]
@chrisduerr okay, that explains the observed behavior when opening a fresh window on `DP-2`. And yes, overriding the scale factor causes the text to be displayed at the desired size/scale; I can set that in my `alacritty.yml` as a way to ensure my text is not scaled when moving between different displays.

This issue is still relevant, though, isn't it? Based on my (now correct) understanding, the `DP-2` to `eDP-1` workflow is not working as intended, as moving from `1.1666...` to `1.6666...` should cause the font to scale up, not down, like it does (slightly).

**Comment by chrisduerr** [comment]
I'd assume your monitor swapping shenanigans are just messing with things. Alacritty might not get informed sufficiently that it actually changed from one monitor to another.

We'd need the `--print-events` logs for checking that though.

**Comment by sudoforge** [comment]
I'll get that uploaded here later today.

**Comment by chrisduerr** [comment]
Running it through `| grep "ScaleFactorChanged"` will give you the useful information.

**Comment by kaimast** [comment]
I can reproduce similar behavior by just changing my screen scaling on GNOME with a single monitor.

It seems like Alacritty only checks the scale factor on startup; at least on X11.

**Comment by sudoforge** [comment]
> I'll get that uploaded here later today.

Whoops, I completely forgot about this. I'll get the `--print-events` log here later tonight.

**Comment by kaimast** [comment]
I'm attaching my log.

I change the scale factor to 2x and then back to 1x. This is on a regular 1080p screen.

Looking at the look I can only find the Resized events, there seems to be nothing about the scale factor changing. 

[alacritty.log](https://github.com/alacritty/alacritty/files/6488128/alacritty.log)

**Comment by mem** [comment]
I'm having a similar issue with fonts that are larger than expected.

I was wondering about these lines in the log, which I also have:

```
[2021-02-15 20:07:24.668708406] [DEBUG] [alacritty] Estimated window size: None
[2021-02-15 20:07:24.789380175] [INFO ] [alacritty] Device pixel ratio: 1.6666666666666667
[2021-02-15 20:07:24.873159137] [INFO ] [alacritty] Width: 800, Height: 600
```

what's that 800 and 600 referring to?

If I use xrandr to query the size, I get:

```
eDP connected primary 1920x1080+0+0 (normal left inverted right x axis y axis) 309mm x 174mm
```

both the pixel resolution and the physical size of the panel are correct.

**Comment by kaimast** [comment]
Seems like winit is still somewhat limited in how it handles (or not) scale factor changes in X11: https://github.com/rust-windowing/winit/issues/1228

**Comment by joe733** [comment]
Here's a GIF capturing the scaling issue:

![DPI Flicker](https://user-images.githubusercontent.com/38415384/185310822-80a1dfb0-b585-4dd4-b9ab-060c65e81ffa.gif)

There's a flicker  while translating b/w monitors.

**Comment by sgon00** [comment]
My environment:

```
OS: Debian 11.4
Version: alacritty 0.11.0
Linux/BSD: X11, KDE (Plasma)
```

I think I have this problem. In my laptop, the default font size looks way too big in the built-in monitor. `-vv` shows `Display scale factor: 1.6666666666666667`. I am new to this area, I don't know why there is a scale factor and everything in my debian and KDE is using built-in default setup.
But when I connect the laptop to an external monitor with the same resolution (both are 1920x1080), the font size looks normal. `-vv` in the external monitor shows `Display scale factor: 1, Font size changed to Size(22) with scale factor of 1`.
In the bult-in monitor, the font is too big to be used, so I have to change the font size to `7`. But font size 7 setup can not be used in external monitor, because that will be way too small.
I never see such weird behavior in any other Linux apps.
I don't really know if the problem is caused by DPI scaling and what DPI scaling really is. As an end user, I don't really care of the definition of DPI. I just want the font size to be consistent between different monitors on one running system.
Thanks a lot.

**Comment by colemickens** [comment]
Am I crazy? Did this not get considerably worse recently?

**Comment by kaimast** [comment]
> Am I crazy? Did this not get considerably worse recently?

I assume there are very few people, if any, working on alacritty or winit that still test on X11.

**Comment by colemickens** [comment]
Maybe I'm bad at reading, and/or need to file a new report. I'm on Sway. Please forgive the noise if I've mistakenly commented on an X11-only-issue, but I'm definitely seeing Alacritty do non-sensical things on Wayland.

**Comment by franzos** [comment]
I'm not sure if this is related; It's happening on Wayland:

```
sway: 1.9
guix with Linux 6.8.5
alacritty 0.13.1
```

When I go from laptop to external monitor, the DPI changes, and alacritty becomes pixelated. This only affects windows that are open before switching - new windows scale just fine. When I switch back, the output looks fine again.

These aren't running side by side, so either laptop or external.
All other application adjust just fine.

![swappy-20240417_091443](https://github.com/alacritty/alacritty/assets/2619225/e3973f8e-d38a-44bc-98f2-5805d6157d3c)

**Comment by kchibisov** [comment]
@franzos unless you can repro on sway-git it's a sway bug. Try also alacritty 0.13.2.

**Comment by kchibisov** [comment]
sway 1.9 has a bug with fractional scaling, sway master shouldn't have it. I'm not sure why they decided to not include a fix, but it's up to them...

Basically your window should be blurry based on the position on the monitor, when fractional scaling is in use.
